packages/common/src/Adaptor.js

```diff
@@ -542,6 +543,93 @@ export function chunk(array, chunkSize) {
   return output;
 }
 
+/**
+ * The function `parseCsv` takes a CSV file string or stream and parsing options as input, and returns a promise that
+ * resolves to the parsed CSV data.
+ * @param {String | Stream} streamOrString - A CSV string or a readable stream
+ * @param {Object} [parsingOptions] - Optional. Parsing options for converting CSV to JSON.
+ *     Possible options:
+ *     - `delimiter` {string|Buffer|[string|Buffer]} - Defines the character(s) used to delimitate the fields inside a record. Default: `','`
+ *     - `quote` {string|Buffer|[string|Buffer]} - Defines the characters used to surround a field. Default: `'"'`
+ *     - `escape` {Buffer|string|null|boolean} - Set the escape character as one character/byte only. Default: `"`
+ *     - `columns` {boolean | array | function} - Generates record in the form of object literals. Default: `true`
+ *     - `bom` {boolean} - Strips the {@link https://en.wikipedia.org/wiki/Byte_order_mark byte order mark (BOM)} from the input string or buffer. Default: `true`
+ *     - `trim` {boolean} - Ignore whitespace characters immediately around the `delimiter`. Default: `true`
+ *     - `ltrim` {boolean} - Ignore whitespace characters from the left side of a CSV field. Default: `true`
+ *     - `rtrim` {boolean} - Ignore whitespace characters from the right side of a CSV field. Default: `true`
+ *     - `chunkSize` {number} - The size of each chunk of CSV data. Default: `Infinity`
+ *     - `skip_empty_lines` {boolean} - The `skip_empty_lines` skips any line which is empty. Default: `true`
+ * @param {function} callback - (Optional) callback function
+ * @returns {Promise} The function returns a Promise that resolves to the result of parsing a CSV `stringOrStream`.
+ */
+export function parseCsv(streamOrString, parsingOptions = {}, callback) {
+  const defaultOptions = {
+    delimiter: ',',
+    quote: '"',
+    escape: '"',
+    columns: true,
+    bom: true,
+    trim: true,
+    ltrim: true,
+    rtrim: true,
+    chunkSize: Infinity,
+    skip_empty_lines: true,
+  };
+
+  const filteredOptions = Object.fromEntries(
+    Object.entries(parsingOptions).filter(([key]) => key in defaultOptions)
+  );
+
+  const options = { ...defaultOptions, ...filteredOptions };
+
+  if (options.chunkSize < 1) {
+    throw new Error('chunkSize must be at least 1');
+  }
+
+  return stateObj => {
+    return new Promise((resolve, reject) => {
+      let state = stateObj;
+      let buffer = [];
+
+      const parser =
+        typeof streamOrString === 'string'
+          ? parse(streamOrString, options)
+          : streamOrString.pipe(parse(options));
+
+      const clearBuffer = () => (buffer = []);
+
+      const buildResponse = state => {
```

Comment:
I have some other comments about `buildResponse` and what it does - but yeah, I also agree that it's not a great name, it's not really a response.

Reply:
@josephjclark i couldn't find other comments about `buildResponse`, Can you tell me more on how we can improve it. 
I have revert the naming to `flushBuffer` now, I used `buildResponse` because it was  using `composeNextState()` which take `state, response` so that kinda made sense in my head. But i am open to a new name that is clear on what the function really does

Reply:
We've sorted this out on github, cheers!